drivers: ethernet: support for numaker m55m1x series

cyliangtw opened the pull request:
Add support for Nuvoton numaker m55m1x series EMAC controller. Also include NOCACHE_MEMORY allocation.

maass-hamburg commented on drivers/ethernet/eth_numaker.c:
```diff
@@ -21,22 +21,38 @@
 /* Device EMAC Interface port */
 #define NUMAKER_GMAC_INTF  0
-/* 2KB Data Flash at 0xFF800 */
+/* 2KB Data Flash */
+#if defined(CONFIG_SOC_M467)
 #define NUMAKER_DATA_FLASH (0xFF800U)
+#elif defined(CONFIG_SOC_M55M1XXX)
+#define NUMAKER_DATA_FLASH (0x2FF800U)
+#else
+#error "SoCs without defined data flash for EMAC!"
+#endif
```
this seems like something, that could be fetched from the device tree.

cyliangtw replied:
Yes, I changed to fetch it from device tree at [49d227d](https://github.com/zephyrproject-rtos/zephyr/commit/49d227dd2d5eeddad33cef6b519d4d4450fdb5b6).

maass-hamburg commented on drivers/ethernet/eth_numaker.c:
```diff
@@ -21,22 +21,33 @@
-/* 2KB Data Flash at 0xFF800 */
-#define NUMAKER_DATA_FLASH (0xFF800U)
+/* 2KB Data Flash */
+#define NUMAKER_DATA_FLASH DT_PROP_OR(DT_NODELABEL(emac), data_flash, 0xFF800U)
```
Can you tell me where in the Technical Reference Manual this address is defined and said that that is for reading a mac address? I unfortunately was unsuccessful to find it.

cyliangtw replied:
No TRM mentioned it.
Just to reserve one data flash for EMAC, system house could burn the specific mac-address in this data flash in factory stage.

maass-hamburg replied:
In zephyr we normally don't allow such undocumented behavior in ethernet drivers. It would be okay if it would be a documented reserved area by the soc manufacturer, but it isn't. In that case, it should be implemented in the app code by the user with using the api like ``int ret = net_mgmt(NET_REQUEST_ETHERNET_SET_MAC_ADDRESS, iface, &params, sizeof(struct ethernet_req_params));`` If I looked correctly in the TRM, I noticed that it was in the area for the application, which I don't thing, as there could be some other data there.

For the m55m1x series this behavior shouldn't be used at all, but for the m46x series if removal or or at least putting a Kconfig option around it, would be the best way.

maass-hamburg replied:
what do the other ethernet collaborators think about it @decsny @lmajewski @pdgendt  ?

cyliangtw replied:
@maass-hamburg ,
Refer to other eth driver implementation, to support gen_random_mac() and removed emac data flash.
Thanks of your advice.

maass-hamburg replied:
good, maybe we should inform existing users via the migration guide that this functionality was removed and they aren't surprised

cyliangtw replied:
hmm,
So far, no user asked how to burn mac-address in this specific data flash in zephyr platform.  Thus, in zephyr platform, there should be no developer to utilize this data flash yet.
Where is the migration guide to inform the change ?

maass-hamburg replied:
Users could have read the source code and figured it out themself.
in doc/releases/migration-guide-4.2.rst

cyliangtw replied:
@maass-hamburg ,
Got it.
The requested change label is still on, is there any other concern ?

maass-hamburg replied:
you didn't pushed the change of the migration-guide

cyliangtw replied:
Thanks and pushed migration-guide-4.2.rst

fabiobaltieri commented on west.yml:
```diff
@@ -198,7 +198,7 @@ manifest:
         - hal
     - name: hal_nuvoton
-      revision: 8eec051270fe45c3f9a1a12de60220e1ab26b579
+      revision: pull/17/head
```
```suggestion
      revision: be1042dc8a96ebe9ea4c5d714f07c617539106d6
```

cyliangtw replied:
Thanks and done.

cyliangtw committed:
drivers: ethernet: support for numaker m55m1x series

Add support for Nuvoton numaker m55m1x series EMAC controller.
Also include NOCACHE_MEMORY allocation.
Support to generate random mac address and remove emac data flash.

Signed-off-by: cyliang tw <cyliang@nuvoton.com>
